test/clippy/BUILD

```diff
@@ -0,0 +1,86 @@
+load(
+    "//rust:rust.bzl",
+    "rust_binary",
+    "rust_clippy",
+    "rust_library",
+    "rust_test",
+)
+
+# Declaration of passing targets.
+
+rust_binary(
+    name = "ok_binary",
+    srcs = ["src/main.rs"],
+    edition = "2018",
+)
+
+rust_library(
+    name = "ok_library",
+    srcs = ["src/lib.rs"],
+    edition = "2018",
+)
+
+rust_test(
+    name = "ok_test",
+    srcs = ["src/lib.rs"],
+    edition = "2018",
+)
+
+# Clippy analysis of passing targets.
+
+rust_clippy(
+    name = "ok_binary_clippy",
+    deps = [":ok_binary"],
+)
+
+rust_clippy(
+    name = "ok_library_clippy",
+    deps = [":ok_library"],
+)
+
+rust_clippy(
+    name = "ok_test_clippy",
+    deps = [":ok_test"],
+    testonly = True,
+)
+
+# Declaration of failing targets.
+
+rust_binary(
+    name = "bad_binary",
+    srcs = ["bad_src/main.rs"],
+    edition = "2018",
+)
+
+rust_library(
+    name = "bad_library",
+    srcs = ["bad_src/lib.rs"],
+    edition = "2018",
+)
+
+rust_test(
+    name = "bad_test",
+    srcs = ["bad_src/lib.rs"],
+    edition = "2018",
+)
+
+# Clippy analysis of failing targets.
+
+rust_clippy(
+    name = "bad_binary_clippy",
+    deps = [":bad_binary"],
+    tags = [ "manual" ],
+)
+
+rust_clippy(
+    name = "bad_library_clippy",
+    deps = [":bad_library"],
+    tags = [ "manual" ],
+)
+
+rust_clippy(
+    name = "bad_test_clippy",
+    deps = [":bad_test"],
+    tags = [ "manual" ],
+    testonly = True,
+)
```

Comment:
These targets all throw errors (from clippy) if they're built (i.e., with `bazel build //test/clippy:bad_binary_clippy`) - hence the "manual" tag.

I *really* want to have a way, in Bazel, to specify something like:

```
expect_build_failure_test(
    name = "binary_failing_under_clippy",
    target = ":bad_binary_clippy",
)
```

Which would try to build the target, and assert that an error was thrown.

I tried following: https://docs.bazel.build/versions/master/skylark/testing.html#failure-testing , to see what could be used here. Unfortunately, the failure seems to be arriving at "execution time" -- the *rules* build fine, it's the execution of Clippy itself which causes an error.

I also tried integrating https://github.com/bazelbuild/bazel-integration-testing , but had some issues, as I couldn't figure out how to reference this parent workspace (or even, if I should).

If ya'll have advice here, it would be really appreciated -- I think I'm at the edge of my Bazel knowledge, but I'd like to do what I can to provide an automated tests!

Reply:
This has always been a hard thing. In bazel we always ended up testing using an external testing framework (e.g. https://github.com/bazelbuild/bazel-integration-testing). I don't think there is a better approach there.